### PR TITLE
innernet: init at 1.0.0

### PR DESCRIPTION
**Pull request by tomberek**
###### Motivation for this change
Trying to package rust application and read https://blog.tonari.no/introducing-innernet

###### Things done
Not sure if bringing in linuxHeaders in this fashion is the right way.
Not sure if LLVM and CLANG options are brought in the right way.

Note: did not test any functionality yet

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/innernet/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  pname = "innernet";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "tonarino";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-BaOeZ/bliakY7K9qibfaQOw9w1aInn5vfAjW8vLzv9w=";
+  };
+  LLVM_CONFIG_PATH = "${lib.getBin llvmPackages.llvm}/bin/llvm-config";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/innernet/default.nix:13]:
```suggestion
  };

  LLVM_CONFIG_PATH = "${lib.getBin llvmPackages.llvm}/bin/llvm-config";
```

File: pkgs/tools/networking/innernet/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  LIBCLANG_PATH = "${llvmPackages.libclang}/lib";
+
+  nativeBuildInputs = with llvmPackages; [ llvm clang linuxHeaders ];
+  buildInputs = [ sqlite ];
+  cargoSha256 = "sha256-9Uv1F/o8gKhigmCp2xjdSCTLpvK66fKfkBQfWzY+KPk=";
+
+  meta = with lib; {
+    description = "A private network system that uses WireGuard under the hood.";
```
Comment (bbigras) [inline pkgs/tools/networking/innernet/default.nix:21]:
```suggestion
    description = "A private network system that uses WireGuard under the hood";
```

>  description
>
>    A short (one-line) description of the package. This is shown by nix-env -q --description and also on the Nixpkgs release pages.
>
>    **Don’t include a period at the end.** Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does. 

https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes

**Comment by bbigras** [comment]
Result of `nixpkgs-review pr 118007` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>innernet</li>
  </ul>
</details>

**Review by 0x4A6F — COMMENTED** [review]
Thanks for your effort!

I've come up with a [similar derivation](https://github.com/0x4A6F/nixpkgs/tree/master-innernet) this morning, because I've forgotten to search for open issues first.

**Comment by alyssais** [comment]
@grahamcofborg build innernet

File: pkgs/tools/networking/innernet/default.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    owner = "tonarino";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-BaOeZ/bliakY7K9qibfaQOw9w1aInn5vfAjW8vLzv9w=";
+  };
+  LLVM_CONFIG_PATH = "${lib.getBin llvmPackages.llvm}/bin/llvm-config";
+  LIBCLANG_PATH = "${llvmPackages.libclang}/lib";
+
+  nativeBuildInputs = with llvmPackages; [ llvm clang linuxHeaders ];
+  buildInputs = [ sqlite ];
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/tools/networking/innernet/default.nix:17]:
```suggestion
  buildInputs = [ sqlite ] ++ lib.optionals stdenv.isDarwin [ Security ];
```

If that is required for darwin.

**Comment by kisik21** [comment]
I'm currently playing around with the package and it looks like at least the `innernet-server` executable works as expected, except that hostnames probably won't resolve anywhere because `innernet` currently edits /etc/hosts (see tonarino/innernet#6). I'm currently compiling innernet for `aarch64-linux` to connect a second peer to my new innernet network and hopefully route some data through!

**Comment by kisik21** [comment]
oh no my kernel on the second host doesn't have wireguard built-in so I guess I'm not gonna test it, but I'm pretty sure that the two commands share most of their internals, judging by the source code, so if one works, the other one surely also works.

**Comment by mcginty** [comment]
FYI, [innernet v1.0.1 is out](https://github.com/tonarino/innernet/releases/tag/v1.0.1) which adds the `--hosts-path [path]` and `--no-write-hosts` options for those who want to customize the hosts behavior.

File: pkgs/tools/networking/innernet/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  cargoSha256 = "sha256-9Uv1F/o8gKhigmCp2xjdSCTLpvK66fKfkBQfWzY+KPk=";
+
+  meta = with lib; {
+    description = "A private network system that uses WireGuard under the hood.";
+    homepage = "https://github.com/tonarino/innernet";
+    license = licenses.mit;
+    maintainers = [ maintainers.tomberek ];
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/tools/networking/innernet/default.nix:24]:
```suggestion
    maintainers = with maintainers; [ _0x4A6F tomberek ];
```

File: pkgs/tools/networking/innernet/default.nix [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{ lib, rustPlatform, fetchFromGitHub, llvmPackages, linuxHeaders, sqlite }:
```
[the remark this answers is not in the document]
Comment (tomberek) [reply pkgs/tools/networking/innernet/default.nix:1]: I guess it's not!

File: pkgs/tools/networking/innernet/default.nix [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitHub {
+    owner = "tonarino";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-OomCSA02ypFVzkYMcmkFREWB6x7oxgpt7x2zRANIDMw=";
+  };
+  LIBCLANG_PATH = "${llvmPackages.libclang}/lib";
+
+  nativeBuildInputs = with llvmPackages; [ llvm clang ];
+  buildInputs = [ sqlite ] ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.Security ];
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/innernet/default.nix:16]:
```suggestion
  buildInputs = [ sqlite ] ++ lib.optionals stdenv.isDarwin [ Security ];
```

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -5460,6 +5460,8 @@ in
 
   infamousPlugins = callPackage ../applications/audio/infamousPlugins { };
 
+  innernet = callPackage ../tools/networking/innernet { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:5463]:
```suggestion
  innernet = callPackage ../tools/networking/innernet {
    inherit (darwin.apple_sdk.frameworks) Security;
  };
```

File: pkgs/tools/networking/innernet/default.nix [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{ lib, stdenv, rustPlatform, fetchFromGitHub, llvmPackages, linuxHeaders, sqlite, darwin ? null }:
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/innernet/default.nix:1]:
```suggestion
{ lib, stdenv, rustPlatform, fetchFromGitHub, llvmPackages, linuxHeaders, sqlite, Security }:
```

**Review by 0x4A6F — APPROVED** [review]

File: pkgs/tools/networking/innernet/default.nix [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    owner = "tonarino";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-OomCSA02ypFVzkYMcmkFREWB6x7oxgpt7x2zRANIDMw=";
+  };
+  LIBCLANG_PATH = "${llvmPackages.libclang}/lib";
+
+  nativeBuildInputs = with llvmPackages; [ llvm clang ];
+  buildInputs = [ sqlite ] ++ lib.optionals stdenv.isDarwin [ Security ];
+  cargoSha256 = "sha256-xNw3IMnWKMlOijohurLoaYmSnqY/+doTm9wDIl5AMxc=";
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/innernet/default.nix:17]:
```suggestion
  cargoSha256 = "sha256-GYNk3j8fjKSo3Qk6Qy0l6kNINK3FxlSYoEkJSx7kVpk=";
```

**Review by 0x4A6F — APPROVED** [review]

**Pull request merged**